### PR TITLE
Avoid unnecessary secret copies in seeded XXH3 hashing

### PR DESCRIPTION
**Pull request by pnn64**
Seeded XXH3 copies the 192-byte default secret even when it can use that secret directly.

## Solution

Reuse the existing constant for inputs up to 240 bytes and for zero seeds. Only prepare a derived secret for longer inputs with a nonzero seed. Hash outputs and public APIs stay the same.

Includes a C-reference test for both hash widths, covering length boundaries, unaligned inputs, and zero/nonzero seeds.

## Benchmarks

Measured on a Xeon E5-2696 v4, Windows x86_64, Rust 1.98.0. Representative 64-byte input with seed `0xdead_beef`, from the first complete run:

| Measurement | Before | After | Change |
|---|---:|---:|---:|
| XXH3-64 time/hash | 17.39 ns | 11.23 ns | -35.4% |
| XXH3-64 throughput | 3.43 GiB/s | 5.31 GiB/s | +54.9% |
| XXH3-64 thread cycles/hash | 38.07 | 24.58 | -35.4% |
| XXH3-128 time/hash | 21.23 ns | 15.76 ns | -25.8% |
| XXH3-128 throughput | 2.81 GiB/s | 3.78 GiB/s | +34.7% |
| XXH3-128 thread cycles/hash | 46.51 | 34.53 | -25.7% |
| Heap allocations/hash | 0 | 0 | Unchanged |

Two complete runs covered 84 cases from 0 bytes to 1 MiB. Each case used 16 alternating old/new pairs in one executable, pinned to one logical CPU. Timing used `Instant`; cycles used Windows `QueryThreadCycleTime`.

Short inputs used approximately 8–47% fewer thread cycles. Larger inputs improved or had statistically inconclusive differences. No wall-time or cycle regressions were detected at the per-case 95% confidence level on this setup.

## Validation

- 30,810 C-reference comparisons and 98,328 old/new comparisons per benchmark run passed.
- Library and property tests, feature checks, Clippy, and formatting passed.
- Additional local allocation and streaming checks passed under Miri.

🤖 AI disclosure: I used Codex with GPT-6 Astra at xhigh reasoning effort to help identify the optimization and construct the benchmark. I have reviewed and understand the resulting implementation.

### CODE REVIEW

**Comment by shepmaster** [comment]
Thanks for this PR! I hadn't previously considered using LLMs for lower-level performance tuning like this, but since you opened this I've been playing around with the idea and it's been enjoyable.

What's wild is how *bad* this commit is on my M1 Macbook. 

<details><summary>Change in times for the small data</summary>
<p>


## xxhash3_64::oneshot
|size|change|
|--:|--:|
|0|+1.2%|
|2|-0.7%|
|6|+2.6%|
|13|-5.5%|
|25|+35.5%|
|50|+44.5%|
|80|+51.6%|
|113|+52.8%|
|135|+49.0%|
|150|+46.9%|
|165|+48.3%|
|185|+37.7%|
|200|+35.7%|
|215|+34.3%|
|230|+31.4%|

## xxhash3_64::oneshot_with_seed
|size|change|
|--:|--:|
|0|-23.6%|
|2|-26.8%|
|6|-33.2%|
|13|-40.8%|
|25|-21.1%|
|50|-2.2%|
|80|+13.1%|
|113|+23.2%|
|135|+9.9%|
|150|+26.0%|
|165|+26.0%|
|185|+21.6%|
|200|+22.3%|
|215|+20.2%|
|230|+18.8%|

## xxhash3_128::oneshot
|size|change|
|--:|--:|
|0|+1.3%|
|2|+0.5%|
|6|+1.1%|
|13|+1.5%|
|25|+21.2%|
|50|+30.7%|
|80|+36.5%|
|113|+43.8%|
|135|+18.2%|
|150|+22.3%|
|165|+23.5%|
|185|+24.8%|
|200|+28.2%|
|215|+29.2%|
|230|+32.1%|

## xxhash3_128::oneshot_with_seed
|size|change|
|--:|--:|
|0|-70.9%|
|2|-35.2%|
|6|-33.9%|
|13|-22.5%|
|25|-8.9%|
|50|+1.9%|
|80|+17.9%|
|113|+24.4%|
|135|+9.6%|
|150|+9.2%|
|165|+14.2%|
|185|+10.9%|
|200|+19.4%|
|215|+18.3%|
|230|+23.3%|

</p>
</details> 

The good news is that I've got some other commits that further improve and ultimately allow the Rust code to beat the C code in my benchmarks on my Mac. I'll work on cleaning up some of my infrastructure changes and rebase this when it's ready, then likely merge as-is.

File: comparison/tests/seeded_oneshot.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#[test]
+fn seeded_oneshot_matches_c_across_short_lengths_and_cutoff() {
```
Comment (shepmaster) [inline comparison/tests/seeded_oneshot.rs:2]: What does this test do that isn't covered by other tests? Specifically, is there something missing in the existing example-based or property-based tests?

**Comment by pnn64** [comment]
Thanks for reviewing my PR! I've really enjoyed looking for micro-optimizations in crates I regularly use in my applications.

Those regressions on small ARM inputs are substantial, especially the impact on the unseeded entry points.. My benchmarks only covered seeded calls on Windows x86_64, so I missed those cases - my bad.

I'm glad you've already found improvements that work well on the Mac - looking forward to see your follow-up changes when they're ready.

File: comparison/tests/seeded_oneshot.rs (hunk 1 above)
Comment (pnn64) [reply comparison/tests/seeded_oneshot.rs:2]: The intent was to guarantee a fixed sweep of lengths, seeds, and offsets on every run, particularly around the 240/241-byte boundary for both seeded variants.

It checks the same agreement with the C implementation as the existing property tests, though. Those already cover offsets and explicitly include zero and `u64::MAX` seeds. I haven't identified a specific bug that this catches which the existing suite misses.

I'm happy to remove the separate test to keep the PR focused, or reduce it to explicit boundary cases in the existing test modules if that would be useful.

**Comment by shepmaster** [comment]
> Representative 64-byte input with seed `0xdead_beef`

I had a chance to run this commit on my x86_64 machine. Generally, `oneshot_with_seed` benefits (sometimes greatly!) while `oneshot` is more of a mixed bag with some big slowdowns, some neutral, and some speedups.

<details><summary>x86_64 performance changes</summary>
<p>

## xxhash3_64
### Tiny data (`oneshot`)

| size| change|
|----:|------:|
|    0| -0.47%|
|    2|  0.43%|
|    6|  1.04%|
|   13|  9.00%|
|   25| 13.69%|
|   50| 16.40%|
|   80| 17.68%|
|  113| 19.88%|
|  135|  5.26%|
|  150|  0.51%|
|  165| -2.56%|
|  185|  3.01%|
|  200| -9.15%|
|  215|  1.65%|
|  230| -3.67%|

### Tiny data (`oneshot_with_seed`)

| size|  change|
|----:|-------:|
|    0| -52.18%|
|    2| -56.23%|
|    6| -58.61%|
|   13| -60.11%|
|   25| -51.55%|
|   50| -44.52%|
|   80| -31.38%|
|  113| -21.63%|
|  135| -20.05%|
|  150| -19.16%|
|  165| -19.44%|
|  185| -19.04%|
|  200| -17.23%|
|  215| -18.21%|
|  230| -15.47%|

## xxhash3_128
### Tiny data (`oneshot`)

| size| change|
|----:|------:|
|    0|  5.36%|
|    2|  6.38%|
|    6| -1.10%|
|   13|  0.94%|
|   25|  2.05%|
|   50|  2.38%|
|   80|  4.18%|
|  113|  2.54%|
|  135|  2.61%|
|  150|  3.75%|
|  165|  0.75%|
|  185| -0.21%|
|  200| -0.06%|
|  215| -0.15%|
|  230|  0.12%|

### Tiny data (`oneshot_with_seed`)

| size|  change|
|----:|-------:|
|    0| -53.69%|
|    2| -48.40%|
|    6| -51.89%|
|   13| -42.35%|
|   25| -37.88%|
|   50| -31.19%|
|   80| -25.89%|
|  113| -22.07%|
|  135| -10.89%|
|  150| -10.33%|
|  165| -13.30%|
|  185| -15.12%|
|  200| -17.29%|
|  215| -13.52%|
|  230| -14.98%|

</p>
</details>

**Comment by shepmaster** [comment]
> AI disclosure

As a request, please make sure that it is you, a real life human, who are talking to me. Don't take my words, put them through a machine in lieu of your brain and copy-paste the output.

File: comparison/tests/seeded_oneshot.rs (hunk 1 above)
Comment (shepmaster) [reply comparison/tests/seeded_oneshot.rs:2]: It *sounds* like you (?) are saying there's no value to this test, which makes me confused as to why it was added. You stated that you reviewed and understand the resulting implementation, so help me understand the value.

What do you mean by 

> reduce it to explicit boundary cases in the existing test modules

Are there some "explicit boundary cases" that the existing test modules don't cover?

If you can't explain it, then it seems like there's a mismatch in "reviewing and understanding" that needs to be hammered out somewhere. And yes, if there's truly no value, then this should be removed as it would be completely redundant with existing tests.

Comment (pnn64) [reply comparison/tests/seeded_oneshot.rs]: The comparison is already covered, but the new test is not completely redundant. By "explicit boundary cases", I mean the seeded 240/241-byte transition. I mean, if we change `<= CUTOFF` to `< CUTOFF`, the existing suite can pass, while this test fails at 240 bytes.

**Comment by pnn64** [comment]
> > Representative 64-byte input with seed `0xdead_beef`
> 
> I had a chance to run this commit on my x86_64 machine. Generally, `oneshot_with_seed` benefits (sometimes greatly!) while `oneshot` is more of a mixed bag with some big slowdowns, some neutral, and some speedups.
> x86_64 performance changes

I gave this a few more shots but was unable to keep the speedups without introducing regressions elsewhere. Given the regressions, feel free to close this PR, unless you'd like to build on the approach.

> > AI disclosure
> 
> As a request, please make sure that it is you, a real life human, who are talking to me. Don't take my words, put them through a machine in lieu of your brain and copy-paste the output.

Fair enough. I occasionally use AI to help investigate and to better convey my own thinking; nevertheless, I'm responsible for what I submit and my replies to you.
